hardware_interface/src/resource_manager.cpp

```diff
@@ -51,116 +81,416 @@ class ResourceStorage
-  void check_hardware_initialization_and_configure(
-    HardwareT & hardware, const HardwareInfo & hardware_info)
-  {
-    if (
-      hardware.initialize(hardware_info).id() !=
-      lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED)
-    {
-      throw std::runtime_error(std::string("Failed to initialize '") + hardware_info.name + "'");
-    }
-    if (hardware.configure().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE)
-    {
-      throw std::runtime_error(std::string("Failed to configure '") + hardware_info.name + "'");
-    }
-  }
-
   template <class HardwareT, class HardwareInterfaceT>
-  void initialize_hardware(
+  void load_hardware(
     const HardwareInfo & hardware_info, pluginlib::ClassLoader<HardwareInterfaceT> & loader,
     std::vector<HardwareT> & container)
   {
+    RCUTILS_LOG_INFO_NAMED(
+      "resource_manager", "Loading hardware '%s' ", hardware_info.name.c_str());
     // hardware_class_type has to match class name in plugin xml description
     // TODO(karsten1987) extract package from hardware_class_type
     // e.g.: <package_vendor>/<system_type>
     auto interface = std::unique_ptr<HardwareInterfaceT>(
       loader.createUnmanagedInstance(hardware_info.hardware_class_type));
     HardwareT hardware(std::move(interface));
     container.emplace_back(std::move(hardware));
-    hardware_status_map_.emplace(
-      std::make_pair(container.back().get_name(), container.back().get_state()));
+    // initialize static data about hardware component to reduce later calls
+    HardwareComponentInfo component_info;
+    component_info.name = hardware_info.name;
+    component_info.type = hardware_info.type;
+    component_info.class_type = hardware_info.hardware_class_type;
+    hardware_info_map_.insert(std::make_pair(component_info.name, component_info));
+  }
+
+  template <class HardwareT>
+  bool initialize_hardware(const HardwareInfo & hardware_info, HardwareT & hardware)
+  {
+    RCUTILS_LOG_INFO_NAMED(
+      "resource_manager", "Initialize hardware '%s' ", hardware_info.name.c_str());
+
+    bool result = hardware.initialize(hardware_info).id() ==
+                  lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED;
+
+    if (result)
+    {
+      RCUTILS_LOG_INFO_NAMED(
+        "resource_manager", "Successful initialization of hardware '%s'",
+        hardware_info.name.c_str());
+    }
+    else
+    {
+      RCUTILS_LOG_INFO_NAMED(
+        "resource_manager", "Failed to initialize hardware '%s'", hardware_info.name.c_str());
+    }
+    return result;
+  }
+
+  template <class HardwareT>
+  bool configure_hardware(HardwareT & hardware)
+  {
+    bool result = trigger_hardware_state_transition(
+      std::bind(&HardwareT::configure, &hardware), "configure", hardware.get_name(),
+      lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE);
+
+    if (result)
+    {
+      // TODO(destogl): is it better to check here if previous state was unconfigured instead of
```

Comment:
if we rely on the lifecycle state machine I think we should take it as contract that if we get a component that is in unconfigured state, we can configure it and theoretically I don't think we'd even need to check for state interfaces however I think that's what people will screw up so checking and giving them an informative exception message instead of a `Segmentation Fault` is important.

Reply:
OK, so we keep it as it is with interface checks? I will then simply delete this TODO

Reply:
I'll create a follow-up ticket to decide what's the best way to throw an informative exception